cpp/src/arrow/python/udf.cc

```diff
@@ -0,0 +1,128 @@
+                          int num_args, Datum* out) {
+  std::shared_ptr<Scalar> c_res_data;
+  PyObject* arg_tuple = PyTuple_New(num_args);
+  for (int arg_id = 0; arg_id < num_args; arg_id++) {
+    if (!batch[arg_id].is_scalar()) {
+      return Status::Invalid("Input type and data type doesn't match");
+    }
+    auto c_data = batch[arg_id].scalar();
+    PyObject* data = wrap_scalar(c_data);
+    PyTuple_SetItem(arg_tuple, arg_id, data);
+  }
+  PyObject* result = PyObject_CallObject(function, arg_tuple);
+  if (result == NULL) {
+    return Status::ExecutionError("Output is null, but expected a scalar");
+  }
+  if (!is_scalar(result)) {
+    return Status::Invalid("Output from function is not a scalar");
+  }
+  ARROW_ASSIGN_OR_RAISE(auto unwrapped_result, unwrap_scalar(result));
+  *out = unwrapped_result;
+  return Status::OK();
+}
+
+Status ExecFunctionArray(const compute::ExecBatch& batch, PyObject* function,
+                         int num_args, Datum* out) {
+  std::shared_ptr<Array> c_res_data;
+  PyObject* arg_tuple = PyTuple_New(num_args);
+  for (int arg_id = 0; arg_id < num_args; arg_id++) {
+    if (!batch[arg_id].is_array()) {
+      return Status::Invalid("Input type and data type doesn't match");
+    }
+    auto c_data = batch[arg_id].make_array();
+    PyObject* data = wrap_array(c_data);
+    PyTuple_SetItem(arg_tuple, arg_id, data);
+  }
+  PyObject* result = PyObject_CallObject(function, arg_tuple);
+  if (result == NULL) {
+    return Status::ExecutionError("Output is null, but expected an array");
+  }
+  if (!is_array(result)) {
+    return Status::Invalid("Output from function is not an array");
+  }
+  return unwrap_array(result).Value(out);
+}
+
+Status ScalarUdfBuilder::MakeFunction(PyObject* function, ScalarUdfOptions* options) {
+  if (function == NULL) {
+    return Status::ExecutionError("python function cannot be null");
+  }
+  Py_INCREF(function);
+  function_.reset(function);
+  if (!PyCallable_Check(function_.obj())) {
+    return Status::TypeError("Expected a callable python object.");
+  }
+  auto doc = options->doc();
+  auto arity = options->arity();
+  scalar_func_ = std::make_shared<compute::ScalarFunction>(options->name(), arity, doc);
+
+  // lambda function
+  auto exec = [this, arity](compute::KernelContext* ctx, const compute::ExecBatch& batch,
+                            Datum* out) -> Status {
+    PyAcquireGIL lock;
+    RETURN_NOT_OK(VerifyArityAndInput(arity, batch));
+    if (batch[0].is_array()) {  // checke 0-th element to select array callable
+      RETURN_NOT_OK(ExecFunctionArray(batch, function_.obj(), arity.num_args, out));
+    } else if (batch[0].is_scalar()) {  // check 0-th element to select scalar callable
+      RETURN_NOT_OK(ExecFunctionScalar(batch, function_.obj(), arity.num_args, out));
+    } else {
+      return Status::Invalid("Unexpected input type, scalar or array type expected.");
+    }
```

Comment:
Ah, and this is about the output type? Shouldn't we look at `options->output_type` to determine that?

Reply:
Thinking about this, I am not sure if I have missed an important point.

Not only as a separate function execution, but when we think of the execution engine, assume if there is a scalar udf taking input as an array but should it only be able to return an array as the output? At the moment we have limited this, such that a scalar function of input array outputs an array and input scalar outputs a scalar. This can be trivially modified to support the output to be a scalar or an array. @westonpace does this relate to the VectorNode support for VectorUDFs too?

Reply:
Ah, I think a scalar function can only be array->array or scalar->scalar, yes. This would be different of course for vector functions. But I think we need to check all of the input shapes then, not just a single one. And I believe we can avoid having two largely identical functions as well.

Reply:
@lidavidm I added a validation for the input types.

Reply:
Hmm, from purely a "can it be used in a project node" standpoint I think array->scalar and scalar->array is fine.  I can't actually come up with any real world examples though.  array -> scalar would most likely be used in some kind of aggregate.

Maybe my best example of scalar -> array would be some function that "uniqifies" a column, guaranteeing each value is unique.  So an array input of `["foo", "bar", "bar"]` would return `["foo", "bar", "bar0"]` and a scalar input of ("foo", length=3) would return `["foo", "foo0", "foo1"]`.

One more thing to consider is that "random" is a fun scalar function that has 0-arity (nullary) and outputs an array.  We should probably be able to support this kind of case too.

Reply:
Ah, I just realized my "uniqify" example is technically not scalar because the output is not based on a single row (although we will have no way of preventing users from writing such a function).  Still, the original point stands, the definition of "scalar function" still allows for functions that are scalar->vector and vector->scalar.

However, as long as we support the random case, I think the other two cases are rare enough we can tackle them in future PRs (and even then, only if a user were able to demonstrate a compelling reason)

Reply:
I think this is a healthy dicussion to create a few more tickets as follow up PRs to expand the supported scope of UDFs. 
And I will include some tests for random-like functions.